### PR TITLE
Add numeric data structure

### PR DESCRIPTION
**Pull request by timotheeguerin**
Adding new `Numeric` data structure for the value world.

This PR doesn't export it so it won't be usable just yet but this sparate this quite complex entity from the rest of the value world PR

### CODE REVIEW

File: packages/compiler/src/core/numeric.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
+interface Numeric {
+  // (value: string): Numeric;
+  /**
+   * Return the value as JavaScript number.
+   * @throws if the value is not representable as number(not representable in a float64/double)
+   */
+  asNumber(): number;
```
Comment (timotheeguerin) [inline packages/compiler/src/core/numeric.ts:7]: change to not throw

File: packages/compiler/src/core/numeric.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  asString(): string;
```
Comment (timotheeguerin) [inline packages/compiler/src/core/numeric.ts:8]: as `asBigInt`

File: packages/compiler/src/core/numeric.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+/** Error emitted when calling `asNumber()` on a numeric that is to large */
+export interface InvalidNumberError extends Error {
+  code: "InvalidNumber";
+}
+
+const InvalidNumbericError = class extends Error {
+  code = "InvalidNumeric";
+};
+const InvalidNumberError = class extends Error {
```
Comment (timotheeguerin) [inline packages/compiler/src/core/numeric.ts:35]: remove

File: packages/compiler/src/core/numeric.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+export interface InvalidNumbericError extends Error {
+  code: "InvalidNumeric";
+}
+
+/** Error emitted when calling `asNumber()` on a numeric that is to large */
+export interface InvalidNumberError extends Error {
+  code: "InvalidNumber";
+}
+
+const InvalidNumbericError = class extends Error {
```
Comment (timotheeguerin) [inline packages/compiler/src/core/numeric.ts:32]: typo

File: packages/compiler/test/core/numeric.test.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
+import { describe, expect, it } from "vitest";
```
Comment (timotheeguerin) [inline packages/compiler/test/core/numeric.test.ts:1]: add test to verify parse invalid number throws

File: packages/compiler/test/core/numeric.test.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+import { Numeric } from "../../src/core/numeric.js";
+
```
Comment (timotheeguerin) [inline packages/compiler/test/core/numeric.test.ts:3]: add test for `asNumber` and `asbigint

File: packages/compiler/src/core/numeric.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+  readonly n: bigint;
+  /** Exponent */
+  readonly e: number;
+  /** Number of decimal digits */
+  readonly d: number;
+  /** Sign */
+  readonly s: 1 | -1;
+}
+
+export interface InvalidNumericError extends Error {
```
Comment (bterlson) [inline packages/compiler/src/core/numeric.ts:32]: Just use a class, it's fine 😂

**Review by bterlson — APPROVED** [review]

**Pull request merged**